Engine/Animation/Tween/Tween.cpp

```diff
@@ -0,0 +1,270 @@
+	tween->desired_vector = end_value;
+	tween->duration = desired_time;
+	tween->tween_type = TweenType::TRANSLATE;
+
+	return tween;
+}
+
+Tween * Tween::LORotate(ComponentTransform2D* transform, float end_value, float desired_time)
+{
+	Tween* tween = new Tween();
+	tween->transform = transform;
+	tween->initial_value = transform->GetGlobalRotation().z;
+	tween->desired_value = end_value;
+	tween->duration = desired_time;
+	tween->tween_type = TweenType::ROTATE;
+
+	return tween;
+}
+
+Tween * Tween::LOScale(ComponentTransform2D* transform, float end_scale, float desired_time)
+{
+	Tween* tween = new Tween();
+	tween->transform = transform;
+	tween->initial_value = 1.0f;
+	tween->desired_value = end_scale;
+	tween->duration = desired_time;
+	tween->tween_type = TweenType::SCALE;
+
+	return tween;
+}
+
+Tween * Tween::LOColor(ComponentImage* image, float4 end_color, float desired_time)
+{
+	Tween* tween = new Tween();
+	tween->image = image;
+	tween->initial_color = image->color;
+	tween->desired_color = end_color;
+	tween->duration = desired_time;
+	tween->tween_type = TweenType::COLOR;
+
+	return tween;
+}
+
+Tween * Tween::SetEase(EaseType ease)
+{
+	ease_type = ease;
+
+	return this;
+}
+
+Tween* Tween::SetLoops(int loops, TweenLoop loop_type)
+{
+	this->loops = loops;
+	this->loop_type = loop_type;
+
+	return this;
+}
+
+void Tween::Update(float dt)
+{
+	if (state == TweenState::COMPLETED || state == TweenState::PAUSED) return;
+
+	current_time += dt / 1000.0f;
+
+	float progress = NormalizedElapsedTime();
+	APP_LOG_INFO("Tween progress: %f", progress);
+
+	UpdateTweenByType();
+
+	CheckFinishTween(progress);
+}
+
+float Tween::UpdateTweenByType()
+{
+	if (tween_type == Tween::NONE) return 0.0f;
+
+	float eased_time = EasedTime();
+
+	switch (tween_type)
+	{
+	case Tween::TRANSLATE:
+		tweened_vector = float2::Lerp(initial_vector, desired_vector, eased_time);
+		transform->SetTranslation(float3(tweened_vector, 0));
+		break;
+
+	case Tween::ROTATE:
+		tweened_value = math::Lerp(initial_value, desired_value, eased_time);
+		transform->SetRotation(float3(0, 0, tweened_value));
+		break;
+
+	case Tween::SCALE:
+		tweened_value = math::Lerp(initial_value, desired_value, eased_time);
+		transform->SetScale(float3::one * tweened_value);
+		break;
+
+	case Tween::COLOR:
+		tweened_color = float4::Lerp(initial_color, desired_color, eased_time);
+		image->SetColor(tweened_color);
+		break;
+	}
+
+	return eased_time;
+}
+
+void Tween::CheckFinishTween(float progress)
+{
+	if (progress >= 1.0f)
+	{
+		played_times += 1;
+		current_time = start_time;
+
+		if (loops != -1 && played_times >= loops)
+		{
+			state = TweenState::COMPLETED;
+		}
+		else if (loop_type == TweenLoop::RESTART)
+		{
+			ResetTween();
+		}
+	}
+}
+
+float Tween::NormalizedElapsedTime()
+{
+	//A = start_time / 0
+	//B = start_time + duration / 1
+
+	//C = current_time / y
+
+	float normalized_time =  (current_time - start_time) * (1.0f / duration);
+
+	return normalized_time;
```

Comment:
This method should be const and you can return directly the line 153

Reply:
Done!